src/portable/portable.c

```diff
@@ -940,11 +940,75 @@ static int make_marker_text(const char *image_path, OrderedHashmap *extension_im
         return 0;
 }
 
+static int append_release_log_fields(
+                const PortableMetadata *release,
+                ImageClass type,
+                const char *field_name,
+                char **ret_text) {
+
+        static const char *const field_versions[_IMAGE_CLASS_MAX][4]= {
+                 [IMAGE_PORTABLE] = { "IMAGE_VERSION", "VERSION_ID", "BUILD_ID", NULL },
+                 [IMAGE_EXTENSION] = { "SYSEXT_IMAGE_VERSION", "SYSEXT_VERSION_ID", "SYSEXT_BUILD_ID", NULL },
+        };
+        static const char *const field_ids[_IMAGE_CLASS_MAX][3]= {
+                 [IMAGE_PORTABLE] = { "IMAGE_ID", "ID", NULL },
+                 [IMAGE_EXTENSION] = { "SYSEXT_IMAGE_ID", "SYSEXT_ID", NULL },
+        };
+        _cleanup_strv_free_ char **fields = NULL;
+        const char *id = NULL, *version = NULL;
+        int r;
+
+        assert(IN_SET(type, IMAGE_PORTABLE, IMAGE_EXTENSION));
+        assert(!strv_isempty((char *const *)field_ids[type]));
+        assert(!strv_isempty((char *const *)field_versions[type]));
+        assert(field_name);
+        assert(ret_text);
+
+        if (!release)
+                return 0; /* Nothing to do. */
+
+        r = load_env_file_pairs_fd(release->fd, release->name, &fields);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse '%s': %m", release->name);
+
+        /* Find an ID first, in order of preference from more specific to less specific: IMAGE_ID -> ID */
+        STRV_FOREACH(k, (char *const *)field_ids[type]) {
+                id = strv_env_pairs_get(fields, *k);
+                if (id)
+                        break;
+        }
+
+        /* Then the version, same logic, prefer the more specific one */
+        STRV_FOREACH(k, (char *const *)field_versions[type]) {
+                version = strv_env_pairs_get(fields, *k);
+                if (version)
+                        break;
+        }
```

Comment:
doesn't matter, but you could make this a helper call: find_first_field() or so that takes two char** arrays, and does this.

Reply:
it's really a corner case of a corner case, let's leave the helper out until there's a separate use case also appears somewhere else

Reply:
well, you already have the same loop twice here. thats why i am suggesting to make it a helper call...

Reply:
added